### PR TITLE
Fix: gossip topic subscribe backpressure

### PR DESCRIPTION
**Pull request by 0xcb9ff9**
# Description

fix issue:

* `Topic` and `*pubsub.Subscription` not closing on exit
*  Every message creates a goroutine

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: consensus/ibft/ibft.go [hunk 1]
```diff
@@ -74,6 +74,7 @@ type Ibft struct {
 	blockchain blockchainInterface // Interface exposed by the blockchain layer
 	executor   *state.Executor     // Reference to the state executor
 	closeCh    chan struct{}       // Channel for closing
+	isClosed   bool
```
Comment (DarianShawn) [inline consensus/ibft/ibft.go]: Use `atomic.Bool` from uber would be more meaniful?

File: network/gossip.go [hunk 2]
```diff
@@ -19,9 +22,11 @@ const (
 type Topic struct {
 	logger hclog.Logger
 
-	topic   *pubsub.Topic
-	typ     reflect.Type
-	closeCh chan struct{}
+	topic *pubsub.Topic
+	typ   reflect.Type
+
+	wg            *sync.WaitGroup
```
Comment (DarianShawn) [inline network/gossip.go]: No need to use pointer here.

File: network/gossip.go [hunk 3]
```diff
@@ -53,32 +58,62 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
+	close(t.unsubscribeCh)
+	t.wg.Wait()
+
+	return t.topic.Close()
+}
+
 func (t *Topic) readLoop(sub *pubsub.Subscription, handler func(obj interface{})) {
 	ctx, cancelFn := context.WithCancel(context.Background())
 
+	var unsubscribe int32 = 0
```
Comment (DarianShawn) [inline network/gossip.go]: Suggest using uber's `atomic.Bool`.

File: network/gossip.go [hunk 4]
```diff
@@ -53,32 +58,62 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
-			obj := t.createObj()
-			if err := proto.Unmarshal(msg.Data, obj); err != nil {
-				t.logger.Error("failed to unmarshal topic", "err", err)
+		obj := t.createObj()
+		if err := proto.Unmarshal(msg.Data, obj); err != nil {
+			t.logger.Error("failed to unmarshal topic", "err", err)
 
-				return
-			}
+			return
```
Comment (DarianShawn) [inline network/gossip.go:114]: What if the message is coming from an attacker?
Refactor should consider more business over pattern.

**Review by DarianShawn — CHANGES_REQUESTED** [review]
Great job.
But I think some of the codes need to be carefully considered.

File: network/gossip.go [hunk 5]
```diff
@@ -53,32 +58,62 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
+
+		close(workqueue)
+
 		cancelFn()
+		sub.Cancel()
+		t.wg.Done()
 	}()
 
-	for {
+	for i := 0; i < runtime.NumCPU(); i++ {
```
Comment (DarianShawn) [inline network/gossip.go]: Would it be any race condition here?

**Comment by 0xcb9ff9** [comment]
@DarianShawn  9e5f22d ok, now use uber's `atomic` library

File: network/gossip.go (hunk 4 above)
Comment (0xcb9ff9) [reply network/gossip.go:114]: If the message comes from an attacker, creating goroutine will run out of memory (not waiting message handler, not drop message, buffer size limit invalid [gossip.go#L19](https://github.com/dogechain-lab/dogechain/blob/c6944f2038cf974cc9c58f5f7efc3ccc1b306461/network/gossip.go#L19))

Comment (DarianShawn) [reply network/gossip.go:114]: e.g. 
An attacker send a not known type message, all nodes would return and do not receive any message any more?

File: network/gossip.go (hunk 5 above)
Comment (0xcb9ff9) [reply network/gossip.go]: general worker pool size will use this, maybe need clamp, too many goroutines can increase CPU contention

File: network/gossip.go (hunk 4 above)
Comment (0xcb9ff9) [reply network/gossip.go:114]: fcfcbc4 I missed this, I should use continue

File: network/gossip.go [hunk 6]
```diff
@@ -53,32 +58,62 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
 	return nil
 }
 
+func (t *Topic) Close() error {
+	close(t.unsubscribeCh)
+	t.wg.Wait()
```
Comment (DarianShawn) [inline network/gossip.go:68]: Is the `waitGroup` really necessary? I would recommend remove this `wg` field.

Comment (0xcb9ff9) [reply network/gossip.go:68]: wait pubsub send an unsubscribe announcement to the network. 
[network/gossip.go#L84](https://github.com/dogechain-lab/dogechain/blob/fcfcbc440426e08bff34b90d851de2b0928d33cb/network/gossip.go#L84)

Comment (0xcb9ff9) [reply network/gossip.go:68]: maybe need time out handle

File: network/gossip.go [hunk 7]
```diff
@@ -53,32 +60,77 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
+		<-t.unsubscribeCh
+		unsubscribe.Store(true)
+
+		close(workqueue)
+
+		// send cancel timeout
+		timeout := time.NewTimer(30 * time.Second)
+		defer timeout.Stop()
+
+		cancelCh := make(chan struct{}, 1)
```
Comment (DarianShawn) [inline network/gossip.go]: No need buffer channel.

File: network/gossip.go [hunk 8]
```diff
@@ -53,32 +60,77 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
+			t.logger.Error("unmarshal message from", "peer", msg.GetFrom())
 
-				return
-			}
+			continue
+		}
 
-			handler(obj)
-		}()
+		workqueue <- obj
```
Comment (DarianShawn) [inline network/gossip.go]: Check if it would crash or not?

**Review by DarianShawn — COMMENTED** [review]
Some more questions.
In fact, use multiple `chan, waitgroup, timer` waiting for each other makes things complicated.
Could we provide a more elegant way doing this?

File: network/gossip.go (hunk 8 above)
Comment (0xcb9ff9) [reply network/gossip.go]: 0e830e2 close channel from sender

File: consensus/ibft/ibft_test.go [hunk 9]
```diff
@@ -1211,6 +1212,7 @@ func newMockIbft(t *testing.T, accounts []string, account string) *mockIbft {
 		validatorKey:     addr.priv,
 		validatorKeyAddr: addr.Address(),
 		closeCh:          make(chan struct{}),
+		isClosed:         atomic.NewBool(false),
```
Comment (DarianShawn) [inline consensus/ibft/ibft_test.go:1215]: Provide more coverage tests to ensure some edge cases please.

File: network/gossip.go [hunk 10]
```diff
@@ -53,33 +63,81 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
 	return nil
 }
 
+func (t *Topic) Close() error {
```
Comment (DarianShawn) [inline network/gossip.go:66]: Unit tests should cover it.
Then we will notice whether it works or it breaks anything.

File: network/gossip.go [hunk 11]
```diff
@@ -53,33 +63,81 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
+	close(t.unsubscribeCh)
+	t.wg.Wait()
+
+	return t.topic.Close()
+}
+
 func (t *Topic) readLoop(sub *pubsub.Subscription, handler func(obj interface{})) {
 	ctx, cancelFn := context.WithCancel(context.Background())
+	workqueue := make(chan proto.Message, workerNum)
```
Comment (DarianShawn) [inline network/gossip.go:77]: Considering `readloop` would not return until stop, why not use `defer close(workqueue)` here?

Comment (0xcb9ff9) [reply network/gossip.go:77]: You are right [7d7e086](https://github.com/dogechain-lab/dogechain/pull/203/commits/7d7e086bdcd25f0b679e73f6d13d9b3672fa031b)

File: network/gossip.go (hunk 10 above)
Comment (0xcb9ff9) [reply network/gossip.go:66]: test added [7d7e086](https://github.com/dogechain-lab/dogechain/pull/203/commits/7d7e086bdcd25f0b679e73f6d13d9b3672fa031b)

File: consensus/ibft/ibft_test.go (hunk 9 above)
Comment (0xcb9ff9) [reply consensus/ibft/ibft_test.go:1215]: this is fix `mockIbft` double close, reused use here 
[consensus/ibft/ibft_test.go#L1236](https://github.com/dogechain-lab/dogechain/blob/9e7a7ab755bac0181422269ce41b7a1162f396f1/consensus/ibft/ibft_test.go#L1236)

File: network/gossip.go (hunk 10 above)
Comment (DarianShawn) [reply network/gossip.go:66]: Great test idea 👍🏻

File: network/gossip.go [hunk 12]
```diff
@@ -53,32 +63,77 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
...
+	}
 
 	for {
-		msg, err := sub.Next(ctx)
-		if err != nil {
-			t.logger.Error("failed to get topic", "err", err)
+		select {
+		case <-t.unsubscribeCh:
+			// send cancel timeout
+			timeout := time.NewTimer(30 * time.Second)
```
Comment (DarianShawn) [inline network/gossip.go]: Should we use `context.WithTimeout()` instead of `timer + chan`?
It is more elegant and bug-free.
Check this out: https://go.dev/play/p/7lDLS9Jfqvz

**Review by DarianShawn — COMMENTED** [review]
Final suggestion for the long term.

File: network/gossip.go (hunk 12 above)
Comment (0xcb9ff9) [reply network/gossip.go]: ok [ac6a06c](https://github.com/dogechain-lab/dogechain/pull/203/commits/ac6a06cdc8a0e7d05fa1bd819e62aafd1b5585a2)

**Review by DarianShawn — APPROVED** [review]
LGTM
